treewide: CUDA reformat with `nixfmt-rfc-style`

ConnorBaker opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Reformats the following paths with `nixfmt-rfc-style`:

- `pkgs/development/cuda-modules`
- `pkgs/test/cuda`
- `pkgs/top-level/cuda-packages.nix`

Additionally, adds the commit in which the reformatting happens to `.git-blame-ignore-revs`.

Ideally this will not cause any rebuilds, but it may if whitespace in raw text blocks change.

This PR also introduces a workflow to ensure that the changed files continue to stay formatted.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
{pkgs/development/cuda-modules,pkgs/test/cuda,pkgs/top-level/cuda-packages.nix}: reformat all CUDA files with nixfmt-rfc-style 2023-03-01

```bash
nix run github:NixOS/nixpkgs/ab6071eb54cc9b66dda436111d4f569e4e56cbf4#nixfmt-rfc-style -L --allow-import-from-derivation -- pkgs/development/cuda-modules pkgs/test/cuda pkgs/top-level/cuda-packages.nix
```

Commit:
git-blame-ignore-revs: ignore CUDA-related reformatting

ConnorBaker commented:
```console
PR="299578"; \
SYSTEM="x86_64-linux"; \
CUDA_SUPPORT="true"; \
CUDA_CAPABILITIES='[ "7.5" ]'; \
nixpkgs-review pr "$PR" \
  --system "$SYSTEM" \
  --no-shell \
  --checkout commit \
  --allow aliases \
  --build-args "--max-jobs 1" \
  --skip-package-regex 'python3(\d){1,2}Packages\.(pytorch-pfn-extras|ffcv)' \
  --skip-package-regex 'python3(\d){1,2}Packages\.(shap|mlflow|optuna)' \
  --extra-nixpkgs-config "{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = ${CUDA_SUPPORT:-false};
    cudaCapabilities = ${CUDA_CAPABILITIES:-[]};
  }"
```

Result of `nixpkgs-review pr 299578 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = true;
    cudaCapabilities = [ "7.5" ];
  }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>cudaPackages.saxpy</li>
    <li>cudaPackagesGoogle.saxpy</li>
    <li>cudaPackages_10.saxpy</li>
  </ul>
</details>

ConnorBaker commented:
```console
PR="299578"; \
SYSTEM="aarch64-darwin"; \
CUDA_SUPPORT="false"; \
CUDA_CAPABILITIES='[ ]'; \
nixpkgs-review pr "$PR" \
  --system "$SYSTEM" \
  --no-shell \
  --checkout commit \
  --allow aliases \
  --build-args "--max-jobs 1" \
  --extra-nixpkgs-config "{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = ${CUDA_SUPPORT:-false};
    cudaCapabilities = ${CUDA_CAPABILITIES:-[]};
  }"
```

Result of `nixpkgs-review pr 299578 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = false;
    cudaCapabilities = [ ];
  }'` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

ConnorBaker commented:
```console
PR="299578"; \
SYSTEM="x86_64-darwin"; \
CUDA_SUPPORT="false"; \
CUDA_CAPABILITIES='[ ]'; \
nixpkgs-review pr "$PR" \
  --system "$SYSTEM" \
  --no-shell \
  --checkout commit \
  --allow aliases \
  --build-args "--max-jobs 1" \
  --extra-nixpkgs-config "{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = ${CUDA_SUPPORT:-false};
    cudaCapabilities = ${CUDA_CAPABILITIES:-[]};
  }"
```

Result of `nixpkgs-review pr 299578 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = false;
    cudaCapabilities = [ ];
  }'` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

ConnorBaker commented:
```console
PR="299578"; \
SYSTEM="aarch64-linux"; \
CUDA_SUPPORT="true"; \
CUDA_CAPABILITIES='[ "7.2" ]'; \
nixpkgs-review pr "$PR" \
  --system "$SYSTEM" \
  --no-shell \
  --checkout commit \
  --allow aliases \
  --build-args "--max-jobs 4" \
  --extra-nixpkgs-config "{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = ${CUDA_SUPPORT:-false};
    cudaCapabilities = ${CUDA_CAPABILITIES:-[]};
  }"
```

Result of `nixpkgs-review pr 299578 --extra-nixpkgs-config '{
    allowUnfree = true;
    allowBroken = false;
    cudaSupport = true;
    cudaCapabilities = [ "7.2" ];
  }'` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

samuela commented:
Thanks for doing this @ConnorBaker ! I'm a fan of all things automated formatting. 

How will we enforce this formatting standard in future PRs? Once things are all pretty and pristine after this PR, what's to stop someone from unknowingly breaking formatting in a subsequent PR?

ConnorBaker commented:
> How will we enforce this formatting standard in future PRs? Once things are all pretty and pristine after this PR, what's to stop someone from unknowingly breaking formatting in a subsequent PR?

I don't know :(

As far as I can tell, we don't have any sort of pre-commit hook setup with Nixpkgs. I suppose I could add a new workflow which checks to make sure that the paths formatted by this PR are unchanged when the formatter is run again, but I'm not sure if that's something which is allowed.

Any ideas?

samuela commented:
> I suppose I could add a new workflow which checks to make sure that the paths formatted by this PR are unchanged when the formatter is run again, but I'm not sure if that's something which is allowed.

A GitHub Actions workflow? Yeah I think that makes sense. IIUC there's a `--verify` option that we can make use of.

Perhaps the https://github.com/nixos/rfcs/pull/166 authors (cc @piegamesde @infinisil) might have input on how to enforce the new formatting standard in CI?

Commit:
workflows/check-nix-format.yml: init

ConnorBaker commented:
@samuela see b6024fa48c4240837f276ca603f8176052c7de75 -- I use the `--check` argument so the command should fail with an error if a file isn't formatted.

samuela commented:
> @samuela see https://github.com/NixOS/nixpkgs/commit/b6024fa48c4240837f276ca603f8176052c7de75 -- I use the --check argument so the command should fail with an error if a file isn't formatted.

Woohoo! Looks great!

infinisil commented on .github/workflows/check-cuda-formatted.yml:
The only thing is that I wouldn't want people to add additional GitHub workflows for each part that starts using the unstable format, so instead I'd say let's imply that the GitHub action is specific to CUDA, and instead name it something generic like `check-nix-formatting.yml`

Also please add an explanatory comment in this file to say that it's not stable yet with a link to https://github.com/NixOS/rfcs/pull/166

ConnorBaker replied:
Resolved in latest

infinisil commented on .github/workflows/check-cuda-formatted.yml:
```diff
@@ -0,0 +1,36 @@
+name: Check that CUDA files are formatted
+
+on:
+  pull_request_target:
+    paths:
+      - pkgs/development/cuda-modules/**.nix
+      - pkgs/test/cuda/**.nix
+      - pkgs/top-level/cuda-packages.nix
```
```suggestion
  pull_request_target:
```

We can then run it for all PRs

SomeoneSerge replied:
Do we want to run the check for all PRs at this point? Maybe we maintain (in the workflow definition) a list of subsystems that explicitly opt-in into enforcing the experimental style

piegamesde replied:
I'd say maybe yes, but also this should probably be factored out into a separate PR then

infinisil reviewed (dismissed):
This is a _bit_ early, because `nixfmt-rfc-style` is not yet stable! We still need to release the first official version, and we planned it to only apply to Nixpkgs after that, see https://github.com/NixOS/nixfmt/issues/153.

That said, I don't see any major problems with code owners manually starting to enforce even the unstable format.

piegamesde commented:
Am I seeing correctly that the reformat is causing a rebuild? If so, would you please investigate this and open a bug report? Nixfmt should never change the code's semantics, and thus never cause any rebuilds.

SomeoneSerge commented on .github/workflows/check-cuda-formatted.yml:
```diff
@@ -0,0 +1,36 @@
+            --run \
+              'nixfmt \
+                --check \
+                pkgs/development/cuda-modules \
+                pkgs/test/cuda \
+                pkgs/top-level/cuda-packages.nix'
```
In line with making the check more generic, I wonder if we should replace this with the list of files that changed between `HEAD` and the merge-base. This way we only "complain" about files brought into scope by PR authors

ConnorBaker replied:
That would be neat, but I don't want to dig too far into the GitHub actions stuff. I firmly believe there should be a community-maintained GitHub action which implements that requirement -- it's super common to only want to run against stuff that has changed! I don't know I've seen that done outside pre-commit hooks, but I haven't look too closely.

dasJ replied:
I was told the same when I wrote the nix-parse workflow, and I didn't want to interact with GH actions either… anyway, this is the solution I was hinted to: https://github.com/NixOS/nixpkgs/blob/master/.github/workflows/nix-parse.yml

infinisil commented:
Oh and with a more generic check, the @NixOS/nix-formatting team should be added as the workflows code owner

gabyx commented:
> > I suppose I could add a new workflow which checks to make sure that the paths formatted by this PR are unchanged when the formatter is run again, but I'm not sure if that's something which is allowed.
> 
> A GitHub Actions workflow? Yeah I think that makes sense. IIUC there's a `--verify` option that we can make use of.
> 
> Perhaps the [NixOS/rfcs#166](https://github.com/NixOS/rfcs/pull/166) authors (cc @piegamesde @infinisil) might have input on how to enforce the new formatting standard in CI?

I would make a GIthub action workflow to just run the changed files in the PR through the formatter and if any `diff` arises after the run -> report to the PR and block it from merging (basically happens directly because this workflow would be unsuccessful).

samuela commented:
> if any diff arises after the run -> report to the PR and block it from merging

Is this not the behavior of `nixfmt --check`?

ConnorBaker commented:
> The only thing is that I wouldn't want people to add additional GitHub workflows for each part that starts using the unstable format, so instead I'd say let's not imply that the GitHub action is specific to CUDA, and instead name it something generic like check-nix-formatting.yml
>
> Also please add an explanatory comment in this file to say that it's not stable yet with a link to https://github.com/NixOS/rfcs/pull/166

@infinisil I've renamed the file and added a comment to make it clear its opt-in.

> Do we want to run the check for all PRs at this point? Maybe we maintain (in the workflow definition) a list of subsystems that explicitly opt-in into enforcing the experimental style

@SomeoneSerge @piegamesde I've relaxed the run condition so it will trigger on all PRs, but have restricted it to just the relevant CUDA files.

> I would make a GIthub action workflow to just run the changed files in the PR through the formatter and if any `diff` arises after the run -> report to the PR and block it from merging (basically happens directly because this workflow would be unsuccessful).

@gabyx @samuela yes, that is what the `--check` flag is for. Though, it won't print the diff itself.

> Am I seeing correctly that the reformat is causing a rebuild? If so, would you please investigate this and open a bug report? Nixfmt should never change the code's semantics, and thus never cause any rebuilds.

@piegamesde @infinisil I've often wondered about those counts, especially for unfree redistributables. If they're calculated naively (e.g., does Hydra have a copy of the package cached?) then it would make sense there are always "rebuilds"; Hydra doesn't build unfree packages (like anything involving CUDA). That the Darwin count is non-zero makes me chuckle, because that's definitely one platform NVIDIA doesn't support, and one which should never be impacted by our changes (unless we, you know, break evaluation).

Is there an easy way for me to check exactly what would need to be rebuilt, or to further investigate this?

---

I see the current PR as the minimal amount of work necessary to format the portion of the code the @NixOS/cuda-maintainers team works on, and to keep it formatted.

I think it would be a good idea to open an issue on the `nixfmt` repo to create a list of requirements for a GitHub Action which runs `nixfmt` on a repository. I think that's a good way to gain further adoption; I know if there were a GitHub action specifically for `nixfmt` that I'd use it -- especially if it provided additional functionality like producing a diff of the changed files, like @gabyx suggested.

What are y'all's thoughts on something like that? I think the changes I've made to the PR (finding paths from environment variables) make it flexible enough for a small group of maintainers to opt-in prior to tree-wide formatting, but I wouldn't want to extend it further.

gabyx commented on .github/workflows/check-nix-format.yml:
```diff
@@ -0,0 +1,50 @@
+            pkgs/top-level/cuda-packages.nix
+        # Iterate over all environment variables beginning with NIX_FMT_PATHS_.
+        run: |
+          for env_var in "${!NIX_FMT_PATHS_@}"; do
```
@ConnorBaker: Sorry if I am mistaken, but looking at this, I was think the following:
Why do we not make nix-shell or `nix-build` out of this, which has the following
- Format the files changed given a Git input to `git diff`, (bruteforce running on the repo)
- On `--check` given report if changes happened. Should run isolated in `/tmp` or so, because it does not alter files...

Is that the idea to implement this in `nixfmt`?

gabyx replied:
Basically we only run `nix run nixfmt#check main...PR-123` or something

ConnorBaker replied:
That would be great functionality to have (I advocated for a GitHub action which does exactly that, maintained by the community), but I think it's outside the scope of this PR -- I want the files I'm working with most waking moments of my life to be formatted.

To be clear, I do think we should have a utility which allows people to do that, but that gets much closer to what `pre-commit` and friends do. I don't want this PR to turn into an RFC proposing such a workflow, which is why I'm content with a simple bash `for` loop.

Does that make sense @gabyx?

gabyx replied:
Jeah sure, this `diffing` thing is only a CI issue, I would not go into `pre-commits`, they serve another matter, despite beeeing overlapping.

Since this is nixpkgs, the best would be at the end to have with the efforts @infinisil did, a tool which people can just run inside nixpkgs. e.g. `nix run .#format main...mybranch`. And as a bonus that will straight run inside this nixpkgs :).

@ConnorBaker : I am totally with you, formatting and having CI/Tooling working for such a big OSS project should be key and nobody wants to fight with stupidities as whitespace changes... (-> blame the tool, not the people) =)

gabyx replied:
The changes you did to the Github workflow, I would generally consider good first solution. We could improve that to make it run directly with `nix` on and `alpine-with-nix` container or such... 
In the end we could add more and more paths to this list, and finally remove it and format the whole repo, when it works. @infinisil might have much more infos on that state etc, and maybe thats already the plan but way differently ;)

infinisil reviewed (approved):
Looking good to me! I opened https://github.com/NixOS/nixfmt/issues/180 to track a generic GitHub workflow file.

We'll discuss this PR in today's [formatting team meeting](https://github.com/NixOS/nixfmt/tree/master/team) and will most likely merge it then :)

infinisil reviewed (approved):
(discussed in the formatting team meeting)

Can confirm that the workflow works (https://github.com/tweag/nixpkgs/pull/88), let's merge!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/formatting-team-meeting-2024-04-02/42658/1

samuela commented:
Woohoo!! thanks everyone -- excited to finally get some auto-formatting CI in nixpkgs!

philiptaron commented:
Thank you all for being the first-out-of-the-gate team to opt in to RFC 166!
